src/app/shared/navigation-focus/navigation-focus.service.ts

```diff
@@ -0,0 +1,84 @@
+import {Injectable, OnDestroy} from '@angular/core';
+import {Event, NavigationEnd, Router} from '@angular/router';
+import {filter, skip} from 'rxjs/operators';
+import {Subscription} from 'rxjs';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class NavigationFocusService implements OnDestroy {
+  private subscriptions = new Subscription();
+  private navigationFocusRequests: HTMLElement[] = [];
+  private skipLinkFocusRequests: HTMLElement[] = [];
+  private skipLinkHref = '';
+  private uid = 0;
+
+  readonly navigationEndEvents = this.router.events
+    .pipe(filter((event: Event): event is NavigationEnd => event instanceof NavigationEnd));
+  readonly softNavigations = this.navigationEndEvents.pipe(skip(1));
+
+  constructor(private router: Router) {
+    this.subscriptions.add(this.softNavigations.subscribe(() => {
+      // focus if url does not have fragment
+      if (!this.router.url.split('#')[1]) {
+        setTimeout(() => {
+          if (this.navigationFocusRequests.length) {
+            this.navigationFocusRequests[this.navigationFocusRequests.length - 1]
+              .focus({preventScroll: true});
+          }
+        }, 100);
+      }
+    }));
+  }
+
+  ngOnDestroy() {
+    this.subscriptions.unsubscribe();
+  }
+
+  requestFocusOnNavigation(el: HTMLElement) {
+    this.navigationFocusRequests.push(el);
+  }
+
+  relinquishFocusOnNavigation(el: HTMLElement) {
+    this.navigationFocusRequests.splice(this.navigationFocusRequests.indexOf(el), 1);
+  }
+
+  requestSkipLinkFocus(el: HTMLElement) {
+    this.skipLinkFocusRequests.push(el);
+    this.setSkipLinkHref(el);
+  }
+
+  relinquishSkipLinkFocus(el: HTMLElement) {
+    this.skipLinkFocusRequests.splice(this.skipLinkFocusRequests.indexOf(el), 1);
+    if (this.skipLinkFocusRequests.length) {
+      const skipLinkFocusTarget = this.skipLinkFocusRequests[this.skipLinkFocusRequests.length - 1];
+      this.setSkipLinkHref(skipLinkFocusTarget);
+    } else {
+      this.skipLinkHref = '';
+    }
+  }
+
+  setSkipLinkHref(el: HTMLElement) {
+    const baseUrl = this.router.url.split('#')[0];
+    if (!el.id) {
+      el.id = `skip-link-target-${this.uid++}`;
```

Comment:
can you move this logic to the directive? that seems like a more natural place to have it